[#6070][#5649] docs(core): add credential vending document (#6071)
apache/gravitino: docs/security/credential-vending.md

### ADLS credentials

#### Azure account key credential

A credential with static Azure storage account name and key.

| Gravitino server catalog properties | Gravitino Iceberg REST server configurations        | Description                                               | Default value | Required | Since Version    |
|-------------------------------------|-----------------------------------------------------|-----------------------------------------------------------|---------------|----------|------------------|
| `credential-providers`              | `gravitino.iceberg-rest.credential-providers`       | `azure-account-key` for Azure account key credential.     | (none)        | Yes      | 0.8.0-incubating |
| `azure-storage-account-name`        | `gravitino.iceberg-rest.azure-storage-account-name` | The static storage account name used to access ADLS data. | (none)        | Yes      | 0.8.0-incubating |
| `azure-storage-account-key`         | `gravitino.iceberg-rest.azure-storage-account-key`  | The static storage account key used to access ADLS data.  | (none)        | Yes      | 0.8.0-incubating |

#### ADLS token credential

> A credential with a subscope privileged token, to use OSS token credential, you should create an Microsoft Entra ID service principal and grant proper privilege to it to access the Azure data.

Author: fixed

|-------------------------------------|---------------------------------------------------|---------------------------------------|-------------------------------------|----------|------------------|
| `credential-providers`              | `gravitino.iceberg-rest.credential-providers`     | `gcs-token` for GCS token credential. | (none)                              | Yes      | 0.7.0-incubating |
| `gcs-credential-file-path`          | `gravitino.iceberg-rest.gcs-credential-file-path` | The location of GCS credential file.  | GCS Application default credential. | No       | 0.7.0-incubating |

Please make sure the credential file is accessible by Gravitino, like using `export GOOGLE_APPLICATION_CREDENTIALS=/xx/application_default_credentials.json` before Gravitino Iceberg REST server is started.

## Custom credentials

Gravitino supports custom credentials, you can implement the `org.apache.gravitino.credential.CredentialProvider` interface to support custom credentials, and place the corresponding jar to the classpath of Iceberg catalog server or hadoop catalog.

## Deployment

> Besides set credential related configuration, please downloading Gravitino cloud jar and place it to the classpath of Iceberg catalog server or hadoop catalog.

Reviewer: This whole page needs a proof reading from an English native reviewer.

```suggestion
Besides setting credentials related configuration, please download Gravitino cloud jar and place it in the classpath of the Iceberg catalog server or Hadoop catalog.
```

Author: updated

Gravitino cloud jar:

- [Gravitino AWS bundle jar](https://mvnrepository.com/artifact/org.apache.gravitino/aws-bundle)
- [Gravitino Aliyun bundle jar](https://mvnrepository.com/artifact/org.apache.gravitino/aliyun-bundle)
- [Gravitino GCP bundle jar](https://mvnrepository.com/artifact/org.apache.gravitino/gcp-bundle)
- [Gravitino Azure bundle jar](https://mvnrepository.com/artifact/org.apache.gravitino/azure-bundle)

The classpath of the server:

- Iceberg REST server: the classpath diffs in different deploy mode, please refer to [Server manage](../iceberg-rest-service.md#server-management) part.
- hadoop catalog: `catalogs/hadoop/libs/`

## Scenarios

> ### Credential vending for Iceberg REST server

Reviewer 2: This section seemed to mix up S3 and ADLS.

Author: Could you point out where mixed ADLS?

Reviewer 2: “My bad, I mistook ‘aws’ for ‘Azure’.”

---
title: "Gravitino credential vending"
slug: /security/credential-vending
keyword: security credential vending
license: "This software is licensed under the Apache License version 2."
---

## Background

Gravitino credential vending is used to generate temporary or static credentials for accessing data. With credential vending, Gravitino provides an unified way to control the access to diverse data sources in different platforms.

### Capabilities

- Supports Gravitino Iceberg REST server.
- Supports Gravitino server, only support Hadoop catalog.
- Supports pluggable credentials with build-in credentials:
  - S3: `S3TokenCredential`, `S3SecretKeyCredential`
  - GCS: `GCSTokenCredential`
  - ADLS: `ADLSTokenCredential`, `AzureAccountKeyCredential`
  - OSS: `OSSTokenCredential`, `OSSSecretKeyCredential`

> - Doesn't support Spark/Trino/Flink connector.

Reviewer: ```suggestion
- No support to Spark/Trino/Flink connector yet.
```

Author: updated

> - No support to Spark/Trino/Flink connector yet.

Author: updated

| `s3-secret-access-key`              | `gravitino.iceberg-rest.s3-secret-access-key`      | The static secret access key used to access S3 data.                                                                                                        | (none)        | Yes      | 0.6.0-incubating |
| `s3-role-arn`                       | `gravitino.iceberg-rest.s3-role-arn`               | The ARN of the role to access the S3 data.                                                                                                                  | (none)        | Yes      | 0.7.0-incubating |
| `s3-external-id`                    | `gravitino.iceberg-rest.s3-external-id`            | The S3 external id to generate token.                                                                                                                       | (none)        | No       | 0.7.0-incubating |
| `s3-token-expire-in-secs`           | `gravitino.iceberg-rest.s3-token-expire-in-secs`   | The S3 session token expire time in secs, it couldn't exceed the max session time of the assumed role.                                                      | 3600          | No       | 0.7.0-incubating |

> | `s3-token-service-endpoint`         | `gravitino.iceberg-rest.s3-token-service-endpoint` | An alternative endpoint of the S3 token service, This could be used with s3-compatible object storage service like MINIO that has a different STS endpoint. | (none)        | No       | 0.8.0-incubating |

Reviewer 3: I think it would be beneficial to describe what kind of dynamic credential and how we support it. For example, we have support the Aamazon STS service and utilize STS token to access S3.

Author: add more details for all token credentials

A credential with static OSS access key id and secret access key.

| Gravitino server catalog properties | Gravitino Iceberg REST server configurations      | Description                                                                   | Default value | Required | Since Version    |
|-------------------------------------|---------------------------------------------------|-------------------------------------------------------------------------------|---------------|----------|------------------|
| `credential-providers`              | `gravitino.iceberg-rest.credential-providers`     | `oss-secret-key` for OSS secret credential.                                   | (none)        | Yes      | 0.8.0-incubating |
| `oss-access-key-id`                 | `gravitino.iceberg-rest.oss-access-key-id`        | The static access key ID used to access OSS data.                             | (none)        | Yes      | 0.7.0-incubating |
| `oss-secret-access-key`             | `gravitino.iceberg-rest.oss-secret-access-key`    | The static secret access key used to access OSS data.                         | (none)        | Yes      | 0.7.0-incubating |

#### OSS token credential

> A credential with a subscope privileged token, to use OSS token credential, you should create an role and grant proper privilege to the role to access the OSS data.

Reviewer: ```suggestion
An OSS token is a token credential with scoped privileges.
To use an OSS token credential, you should create a role and grant it proper privileges.
```

Author: updated

|-------------------------------------|---------------------------------------------------|------------------------------------------------------------|-------------------------------------|----------|------------------|
| `credential-providers`              | `gravitino.iceberg-rest.credential-providers`     | `gcs-token` for GCS token credential.                      | (none)                              | Yes      | 0.8.0-incubating |
| `gcs-credential-file-path`          | `gravitino.iceberg-rest.gcs-credential-file-path` | Deprecated, please use `gcs-service-account-file` instead. | GCS Application default credential. | No       | 0.7.0-incubating |
| `gcs-service-account-file`          | `gravitino.iceberg-rest.gcs-service-account-file` | The location of GCS credential file.                       | GCS Application default credential. | No       | 0.8.0-incubating |

:::note
For Gravitino Iceberg REST server, please make sure the credential file is accessible by Gravitino, like using `export GOOGLE_APPLICATION_CREDENTIALS=/xx/application_default_credentials.json` even `gcs-service-account-file` is setting.

> :::

Reviewer: Please tweak this sentence to make sense.

Author: updated

## Custom credentials

Gravitino supports custom credentials, you can implement the `org.apache.gravitino.credential.CredentialProvider` interface to support custom credentials, and place the corresponding jar to the classpath of Iceberg catalog server or Hadoop catalog.

## Deployment

Besides setting credentials related configuration, please download Gravitino cloud bundle jar and place it in the classpath of Iceberg catalog server or Hadoop catalog.

Gravitino cloud bundle jar:

> - [Gravitino AWS bundle jar](https://mvnrepository.com/artifact/org.apache.gravitino/aws-bundle)

Reviewer 3: In 0.8.0, the artifact has the prefix `gravitino`, so the URL should be `https://mvnrepository.com/artifact/org.apache.gravitino/gravitino-aws-bundle`

Author: updated

> Besides setting credentials related configuration, please download Gravitino cloud bundle jar and place it in the classpath of Iceberg catalog server or Hadoop catalog.

Author: updated

- [Gravitino Aliyun bundle jar](https://mvnrepository.com/artifact/org.apache.gravitino/aliyun-bundle)
- [Gravitino GCP bundle jar](https://mvnrepository.com/artifact/org.apache.gravitino/gcp-bundle)
- [Gravitino Azure bundle jar](https://mvnrepository.com/artifact/org.apache.gravitino/azure-bundle)

The classpath of the server:

> - Iceberg REST server: the classpath diffs in different deploy mode, please refer to [Server manage](../iceberg-rest-service.md#server-management) part.

Reviewer 3: It's not advised to abbreviation `diffs` as `differ` is not specialized vocabulary

Author: updated

- Hadoop catalog: `catalogs/hadoop/libs/`

## Scenarios

### Credential vending for Iceberg REST server

Suppose the Iceberg table data is stored in S3, follow the steps below:

> 1. Download the Gravitino AWS bundle jar, and place it to the classpath of Iceberg REST server.

Reviewer 3: Add a link here.

Author: done

> - Iceberg REST server: the classpath differs in different deploy mode, please refer to [Server manage](../iceberg-rest-service.md#server-management) part.

Reviewer 4: "server management"

Author: done

> ## Scenarios

Reviewer 4: `## Usage example`

Author: updated
